File: Halachic Times/app/src/main/java/com/github/times/CandlesFragment.java
/*
 * Copyright 2012, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.times;

import static com.github.times.ZmanimDays.SHABBATH;
import static com.github.times.ZmanimPopulater.BEFORE_SUNSET;
import static com.kosherjava.zmanim.hebrewcalendar.JewishCalendar.CHANUKAH;
import static com.kosherjava.zmanim.hebrewcalendar.JewishCalendar.EREV_YOM_KIPPUR;
import static com.kosherjava.zmanim.hebrewcalendar.JewishCalendar.YOM_KIPPUR;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.github.times.databinding.CandlesChannukaBinding;
import com.github.times.databinding.CandlesFragmentBinding;
import com.github.times.databinding.CandlesKippurimBinding;
import com.github.times.databinding.CandlesShabbatBinding;
import com.github.times.location.ZmanimLocations;
import com.github.times.preference.SimpleZmanimPreferences;
import com.github.times.preference.ZmanimPreferences;
import com.kosherjava.zmanim.util.GeoLocation;

import java.util.Calendar;

/**
 * Shows candle images.
 *
 * @author Moshe Waisberg
 */
public class CandlesFragment extends Fragment {

    /**
     * Provider for locations.
     */
    private ZmanimLocations locations;
    /**
     * The preferences.
     */
    private ZmanimPreferences preferences;

    private CandlesFragmentBinding _binding = null;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        final Context context = requireContext();
        if (context instanceof ZmanimActivity) {
            preferences = ((ZmanimActivity) context).getZmanimPreferences();
        } else {
            preferences = new SimpleZmanimPreferences(context);
        }
        ZmanimApplication app = (ZmanimApplication) context.getApplicationContext();
        locations = app.getLocations();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        CandlesFragmentBinding binding = CandlesFragmentBinding.inflate(inflater, container, false);
        _binding = binding;
        return binding.getRoot();
    }

    @NonNull
    private CandlesPopulater createPopulater(@NonNull Context context) {
        return new CandlesPopulater(context, preferences);
    }

    public void populateTimes(@NonNull Calendar date) {
        // Called before attached to activity?
        if (!isAdded()) return;
        CandlesFragmentBinding binding = _binding;
        if (binding == null) return;

        ZmanimLocations locations = this.locations;
        if (locations == null) return;
        GeoLocation gloc = locations.getGeoLocation();
        // Have we been destroyed?
        if (gloc == null) return;
        Context context = binding.getRoot().getContext();

        CandlesPopulater populater = createPopulater(context);
        populater.setCalendar(date);
        populater.setGeoLocation(gloc);
        populater.setInIsrael(locations.isInIsrael());

        final CandleData candles = populater.populateCandles(preferences);

        final boolean animate = preferences.isCandlesAnimated();
        bindViews(binding, candles, animate);
    }

    private void bindViews(@NonNull CandlesFragmentBinding binding, @NonNull CandleData candles, boolean animate) {
        ViewGroup container = binding.getRoot();
        container.removeAllViews();

        final int holiday = getCandlesHoliday(candles);
        final int candlesCount = candles.countTomorrow;
        CandleView[] candleViews = null;

        Context context = container.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);

        switch (holiday) {
            case EREV_YOM_KIPPUR:
            case YOM_KIPPUR:
                CandlesKippurimBinding bindingKippur = CandlesKippurimBinding.inflate(inflater, container, true);

                candleViews = new CandleView[1];
                candleViews[0] = bindingKippur.candle1;
                break;
            case CHANUKAH:
                CandlesChannukaBinding bindingChannuka = CandlesChannukaBinding.inflate(inflater, container, true);

                candleViews = new CandleView[9];
                candleViews[0] = bindingChannuka.candle1;
                candleViews[1] = bindingChannuka.candle2;
                candleViews[2] = bindingChannuka.candle3;
                candleViews[3] = bindingChannuka.candle4;
                candleViews[4] = bindingChannuka.candle5;
                candleViews[5] = bindingChannuka.candle6;
                candleViews[6] = bindingChannuka.candle7;
                candleViews[7] = bindingChannuka.candle8;
                candleViews[8] = bindingChannuka.candleShamash;

                // Only show relevant candles.
                for (int i = 0; i < candlesCount; i++) {
                    candleViews[i].setVisibility(View.VISIBLE);
                }
                for (int i = candlesCount; i < 8; i++) {
                    candleViews[i].setVisibility(View.INVISIBLE);
                }
                break;
            default:
                if (candlesCount > 0) {
                    CandlesShabbatBinding bindingShabbat = CandlesShabbatBinding.inflate(inflater, container, true);

                    candleViews = new CandleView[2];
                    candleViews[0] = bindingShabbat.candle1;
                    candleViews[1] = bindingShabbat.candle2;
                }
                break;
        }

        setFlickers(candleViews, animate);
    }

    private void setFlickers(@Nullable CandleView[] candles, boolean enabled) {
        if (candles == null)
            return;
        for (CandleView candle : candles) {
            if (candle == null)
                continue;
            candle.flicker(enabled);
        }
    }

    /**
     * Set the view's visibility.
     *
     * @param visibility the visibility.
     * @see View#VISIBLE
     * @see View#INVISIBLE
     * @see View#GONE
     */
    public void setVisibility(int visibility) {
        View view = getView();
        if (view != null) {
            view.setVisibility(visibility);
        }
    }

    public int getCandlesCount() {
        ViewGroup root = (ViewGroup) getView();
        if (root == null) return 0;
        return root.getChildCount();
    }

    /**
     * Get the occasion for lighting candles.
     *
     * @return the candles holiday.
     */
    public int getCandlesHoliday(@NonNull CandleData candles) {
        final int when = candles.whenCandles;
        final int holidayToday = candles.holidayToday;
        final int holidayTomorrow = candles.holidayTomorrow;
        return (when == BEFORE_SUNSET) ? ((holidayTomorrow == SHABBATH) ? holidayTomorrow : holidayToday) : holidayTomorrow;
    }
}

File: Halachic Times/app/src/main/java/com/github/times/CandlesFragment.kt
/*
 * Copyright 2012, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.times

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import com.github.times.ZmanimDays.SHABBATH
import com.github.times.databinding.CandlesChannukaBinding
import com.github.times.databinding.CandlesFragmentBinding
import com.github.times.databinding.CandlesKippurimBinding
import com.github.times.databinding.CandlesShabbatBinding
import com.github.times.location.ZmanimLocations
import com.github.times.preference.SimpleZmanimPreferences
import com.github.times.preference.ZmanimPreferences
import com.kosherjava.zmanim.hebrewcalendar.JewishCalendar
import java.util.Calendar

/**
 * Shows candle images.
 *
 * @author Moshe Waisberg
 */
class CandlesFragment : Fragment() {
    /**
     * Provider for locations.
     */
    private val locations: ZmanimLocations
        get() {
            val app = requireContext().applicationContext as ZmanimApplication
            return app.locations
        }

    /**
     * The preferences.
     */
    private val preferences: ZmanimPreferences by lazy {
        val context = requireContext()
        if (context is ZmanimActivity) {
            context.preferences
        } else {
            SimpleZmanimPreferences(context)
        }
    }

    private var _binding: CandlesFragmentBinding? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val binding = CandlesFragmentBinding.inflate(inflater, container, false)
        _binding = binding
        return binding.root
    }

    private fun createPopulater(context: Context): CandlesPopulater {
        return CandlesPopulater(context, preferences)
    }

    fun populateTimes(date: Calendar) {
        // Called before attached to activity?
        if (!isAdded) return
        val binding = _binding ?: return
        val locations = locations
        val gloc = locations.geoLocation ?: return
        // Have we been destroyed?
        val context = binding.root.context
        val populater = createPopulater(context).apply {
            setCalendar(date)
            setGeoLocation(gloc)
            isInIsrael = locations.isInIsrael
        }
        val candles = populater.populateCandles(preferences) ?: return
        val animate = preferences.isCandlesAnimated
        bindViews(binding, candles, animate)
    }

    private fun bindViews(binding: CandlesFragmentBinding, candles: CandleData, animate: Boolean) {
        val container: ViewGroup = binding.root
        container.removeAllViews()
        val context = container.context
        val inflater = LayoutInflater.from(context)

        val holiday = getCandlesHoliday(candles)
        val candlesCount = candles.countTomorrow
        var candleViews: Array<CandleView>? = null
        when (holiday) {
            JewishCalendar.EREV_YOM_KIPPUR, JewishCalendar.YOM_KIPPUR -> {
                val bindingKippur = CandlesKippurimBinding.inflate(inflater, container, true)
                candleViews = arrayOf(bindingKippur.candle1)
            }

            JewishCalendar.CHANUKAH -> {
                val bindingChannuka = CandlesChannukaBinding.inflate(inflater, container, true)
                candleViews = arrayOf(
                    bindingChannuka.candle1,
                    bindingChannuka.candle2,
                    bindingChannuka.candle3,
                    bindingChannuka.candle4,
                    bindingChannuka.candle5,
                    bindingChannuka.candle6,
                    bindingChannuka.candle7,
                    bindingChannuka.candle8,
                    bindingChannuka.candleShamash
                )

                // Only show relevant candles.
                run {
                    var i = 0
                    while (i < candlesCount) {
                        candleViews!![i].visibility = View.VISIBLE
                        i++
                    }
                }
                var i = candlesCount
                while (i < 8) {
                    candleViews[i].visibility = View.INVISIBLE
                    i++
                }
            }

            else -> if (candlesCount > 0) {
                val bindingShabbat = CandlesShabbatBinding.inflate(inflater, container, true)
                candleViews = arrayOf(
                    bindingShabbat.candle1,
                    bindingShabbat.candle2
                )
            }
        }
        setFlickers(candleViews, animate)
    }

    private fun setFlickers(candles: Array<CandleView>?, enabled: Boolean) {
        if (candles == null) return
        for (candle in candles) {
            candle.flicker(enabled)
        }
    }

    /**
     * Set the view's visibility.
     *
     * @param visibility the visibility.
     */
    fun setVisibility(visibility: Boolean) {
        val view = view
        if (view != null) {
            view.isVisible = visibility
        }
    }

    val candlesCount: Int
        get() {
            val root = view as? ViewGroup ?: return 0
            return root.childCount
        }

    /**
     * Get the occasion for lighting candles.
     *
     * @return the candles holiday.
     */
    fun getCandlesHoliday(candles: CandleData): Int {
        val at = candles.whenCandles
        val holidayToday = candles.holidayToday
        val holidayTomorrow = candles.holidayTomorrow
        return if (at == CandleData.BEFORE_SUNSET) (if (holidayTomorrow == SHABBATH) holidayTomorrow else holidayToday) else holidayTomorrow
    }
}